Murloc7 s5 b alex subscribtion to users bjs2 40508

alexFromAC143937 opened the pull request:
Task BJS2-40508 + тесты
Как пользователь, я хочу иметь возможность подписываться на других пользователей, чтобы сохранить их контакт для дальнейшей совместной работы.

Необходимо реализовать основные операции: подписаться на пользователя, отписаться от пользователя, получить всех подписчиков, получить количество подписчиков, просмотреть свои подписки и получить количество своих подписок. 

Ниже условия для реализации каждой операции. Делаем их поочередно. Можно делать их разными коммитами и отправлять на ревью каждую операцию отдельно, а можно все сразу.

alexFromAC143937 committed:
Task BJS2-40508
1) Controllers - were created;
2) Service Methods - were created;
3) Necessary utils - were created;
4) Tests - were created;
5) Final tests - were done via Postman.

alexFromAC143937 committed:
Task BJS2-40508 + тесты
Как пользователь, я хочу иметь возможность подписываться на других пользователей, чтобы сохранить их контакт для дальнейшей совместной работы.

Необходимо реализовать основные операции: подписаться на пользователя, отписаться от пользователя, получить всех подписчиков, получить количество подписчиков, просмотреть свои подписки и получить количество своих подписок.

Ниже условия для реализации каждой операции. Делаем их поочередно. Можно делать их разными коммитами и отправлять на ревью каждую операцию отдельно, а можно все сразу.

alexFromAC143937 committed:
Task BJS2-40508 + тесты
Решение конфликтов

alexFromAC143937 committed:
Task BJS2-40508 + тесты
Решение конфликтов

alexFromAC143937 committed:
Merge branch 'Murloc-master-stream7' into Murloc7-s5-BAlex-subscribtion-to-users-BJS2-40508

alexFromAC143937 committed:
Task BJS2-40508 + тесты
Рефакторинг наименований

alexFromAC143937 committed:
Task BJS2-40508 + тесты
Рефакторинг наименований

Alex15265 commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,62 @@
+public class SubscriptionController {
+    private final SubscriptionService subscriptionService;
+
+    @Transactional
```
Использование аннотации `@Transactional` на контроллере в некоторых случаях может быть оправданным, но обычно это не считается хорошей практикой.

Аннотация `@Transactional` используется для обозначения методов, которые должны выполняться в рамках транзакции базы данных. В контексте контроллера она может привести к неожиданным результатам, особенно если вы используете несколько методов в одном запросе.

Контроллер обычно отвечает за обработку запросов и передачу их на уровень бизнес-логики. Если вы поместите `@Transactional` на контроллер, то все методы в этом контроллере будут выполняться в одной транзакции. Это может вызвать проблемы с производительностью и масштабируемостью, а также усложнить отладку и тестирование.

Вместо этого рекомендуется использовать `@Transactional` на уровне бизнес-логики (сервисах или репозиториях), где это имеет больше смысла. Контроллер же должен просто вызывать эти методы и передавать данные между клиентом и сервером.

Alex15265 replied:
СhatGpt не рекомендует)

Alex15265 commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Transactional
+    @PostMapping(UrlServiceParameters.FOLLOWING_ADD + UrlServiceParameters.FOLLOWING_PARAMETERS)
+    @ResponseStatus(HttpStatus.OK)
```
как будто лишняя
спринг и так в случае успешного запроса вернет ResponseEntity с 200 статусом

Alex15265 commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,62 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping(UrlServiceParameters.FOLLOWING_SERVICE_URL)
```
Я уже заливал класс UrlUtils.  Лучше в одном месте все константы держать для urlов. 
Ну и к одному виду привести их.
Вроде как приняли такой: api/user-service/v1/.... итд

Alex15265 commented on src/main/java/school/faang/user_service/extention/DataValidationException.java:
```diff
@@ -0,0 +1,7 @@
+package school.faang.user_service.extention;
+
+public class DataValidationException extends RuntimeException {
```
Уже существует готовая ValidationException, можно было ее использовать)
ну эт такое, субъективное мнение)

Alex15265 commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,16 @@
+
+import java.util.List;
+
+public interface SubscriptionService {
```
тут интересно бы Женю послушать)
много раз видел такие промежуточные интерфейсы для сервисов, репо, контроллеров даже
всегда был такой вопрос, нужны ли они вообще, если по факту у них всегда одна имплементация будет
как в данном случае: вряд ли будет несколько SubscriptionServiceImpl

jibulani replied:
да, тоже часто такое встречал, на мой взгляд полезно делать так по нескольким причинам:
1. соблюдаем D из SOLID - dependency inversion - с наличием интерфейса можем делать зависимости от абстракции, а не от реализации
2. если вдруг все-таки понадобится иметь несколько реализаций, то с готовым интерфейсом это проще сделать
3. при наличии интерфейса спринг скорее всего будет использовать dynamic proxy для работы всяких штук типа `@Transactional`, а это немного производительнее, чем cglib

alexFromAC143937 committed:
Task BJS2-40508 + тесты
1) @Transaction - перенес в сервис, перенос в репозиторий - решил не делать, так как если в текущей реализации возможно, то в другом методе, может быть другая необходимость.
2) Url - перенес
3) Exception - Оставил, так как это требование задачи
4) @ResponseStatus(HttpStatus.OK) - удалил

jibulani commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,54 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping(UrlUtils.FOLLOWING_SERVICE_URL)
```
давай приведем к такому же виду, как в других частях проекта: UrlUtils.MAIN_URL + UrlUtils.V1 + UrlUtils.FOLLOWING

jibulani commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,54 @@
+public class SubscriptionController {
+    private final SubscriptionService subscriptionService;
+
+    @PostMapping(UrlUtils.FOLLOWING_ADD + UrlUtils.FOLLOWING_PARAMETERS)
```
UrlUtils.FOLLOWING_ADD  не нужен, сам метод Post уже говорит, что будет что-то создаваться

если ты хочешь, чтобы followerId и followeeId были `@PathVariable`, то FOLLOWING_PARAMETERS должен быть таким: "follower/{followerId}/followee/{followeeId}"
твой текущий FOLLOWING_PARAMETERS больше похож на `@RequestParam`, но тогда для него вообще не нужен FOLLOWING_PARAMETERS

jibulani commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,54 @@
+
+    @PostMapping(UrlUtils.FOLLOWING_ADD + UrlUtils.FOLLOWING_PARAMETERS)
+    public void followUser(@PathVariable long followerId, @PathVariable long followeeId) {
+        if (followerId != followeeId) {
```
эта проверка дважды используется, лучше вынеси ее в отдельный метод валидации, в котором будешь кидать ошибку при совпадении followerId и followeeId, это позволит переиспользовать данный метод, а код станет выглядеть проще, примерно так:
```
validateFollowerAndFallowee(...)
subscriptionService,followUser(...)
```

jibulani commented on src/main/java/school/faang/user_service/extention/DataValidationException.java:
```diff
@@ -0,0 +1,7 @@
+package school.faang.user_service.extention;
```
почему пакет extention называется? имелось в виду exception?

jibulani commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    @PostMapping(UrlUtils.FOLLOWING_DELETE + UrlUtils.FOLLOWING_PARAMETERS)
```
UrlUtils.FOLLOWING_DELETE лишний, чтобы указать, что что-то удаляется лучше использовать `@DeleteMapping`

аналогичная ситуация с FOLLOWING_PARAMETERS, что и в followUser

jibulani commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    @GetMapping(UrlUtils.FOLLOWING_FILTER + "{followeeId}")
```
та же проблема с FOLLOWING_FILTER
лучше дать ему значение "/followers/{followeeId}"

jibulani commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,54 @@
+        return subscriptionService.getFollowers(followeeId, filter);
+    }
+
+    @GetMapping(UrlUtils.FOLLOWING_COUNT + "{followerId}")
```
та же проблема с FOLLOWING_COUNT 
можно дать ему значение "/followers/{followerId}/count"

jibulani commented on src/main/java/school/faang/user_service/utilities/ServiceMethods.java:
```diff
@@ -0,0 +1,13 @@
+import java.sql.Timestamp;
+import java.time.format.DateTimeFormatter;
+
+public class ServiceMethods {
```
лучше назвать класс TimeUtils - тогда будет понятно, в чем именно помогает класс и какими методами его логично расширять, а с текущим названием класс со временем рискует разрастись до огромных размеров, потому что в него будут добавлять все подряд

jibulani commented on src/main/java/school/faang/user_service/utilities/UrlUtils.java:
```diff
@@ -14,4 +14,13 @@ public class UrlUtils {
     public static final String PARTICIPANTS = "/participants";
     public static final String AMOUNT = "/amount";
+
+    public static final String SYSTEM_ID = "001-UserService";
```
что за SYSTEM_ID? где используется и зачем?

jibulani commented on src/test/java/school/faang/user_service/controller/SubscriptionControllerTest.java:
```diff
@@ -0,0 +1,129 @@
+
+    @BeforeEach
+    void setUp() {
+        MockitoAnnotations.openMocks(this);
```
обычно вместо этого над классом добавляют аннотацию `@ExtendWith(MockitoExtension.class)`

jibulani commented on src/test/java/school/faang/user_service/controller/SubscriptionControllerTest.java:
```diff
@@ -0,0 +1,129 @@
+    @Test
+    void followUserSuccess() throws Exception {
+        doNothing().when(subscriptionService).followUser(TEST_ID_USER1, TEST_ID_USER2);
+        System.out.println(UrlUtils.FOLLOWING_SERVICE_URL +
```
не надо в тестах ничего печатать дополнительно
.andDo(print()) должно быть достаточно

jibulani commented on src/test/java/school/faang/user_service/handler/ResponseExceptionHandlerTest.java:
```diff
@@ -0,0 +1,38 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+class ResponseExceptionHandlerTest {
```
не надо такие тесты отдельно выносить, правильная обработка ошибочных сценариев должна быть в тестах на контроллер

jibulani commented on src/test/java/school/faang/user_service/service/impl/SubscriptionServiceImplTest.java:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    @Test
+    void getFollowersSetAllFilterValueSuccessTest() {
```
можно подобные тесты объединить в один с помощью `@ParametrizedTest`

jibulani commented on src/test/java/school/faang/user_service/utilities/ServiceMethodsTest.java:
```diff
@@ -0,0 +1,12 @@
+class ServiceMethodsTest {
+    @Test
+    void getTimeIsoOffsetDateTimeSuccessTest() {
+        assertNotNull(ServiceMethods.getTimeIsoOffsetDateTime(), "ESM001 - Expected a String.");
```
можно еще формат результата работы метода проверить с помощью регулярки, например

jibulani reviewed (changes requested):
есть, что исправить + надо исправить конфликты, а то не получится смержить
хорошо покрыл код тестами, молодец!

alexFromAC143937 committed:
Task BJS2-40508 + тесты
Fix

alexFromAC143937 commented:
![image](https://github.com/user-attachments/assets/56d7d57a-705f-4263-bfcf-19ae8f43f8f4)

alexFromAC143937 commented:
![image](https://github.com/user-attachments/assets/640f1ec9-5f5a-4a04-a51a-29be9b9e83b7)

jibulani reviewed (approved):
все хорошо, осталось поправить конфликты и можно мержить

alexFromAC143937 committed:
Task BJS2-40508
Fix

alexFromAC143937 committed:
Task BJS2-40508
Fix
